### PR TITLE
[REFACTOR][TIR] Migrate most of low-level build to use the Pass Manager.

### PR DESCRIPTION
**Pull request by tqchen**
- SplitHostDevice
- ThreadSync
- BindDevice
- LowerThreadAllreduce
- Provide a temp fix for printing IRModule with PrimFunc before the formal text printer.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @zhiics @yzhliu @ZihengJiang

File: python/tvm/tir/transform/transform.py [hunk 1]
```diff
@@ -18,6 +18,28 @@
 # pylint: disable=invalid-name
 from . import _ffi_api
 
+def BindDeviceType():
+    """Bind the device type ofthe function to be
```
Comment (zhiics) [inline python/tvm/tir/transform/transform.py:22]: of the

File: src/tir/transforms/bind_device_type.cc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#include <tvm/tir/transform.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/analysis.h>
+#include <tvm/target/target.h>
+#include <tvm/runtime/registry.h>
+
+namespace tvm {
+namespace tir {
+
+
```
Comment (zhiics) [inline src/tir/transforms/bind_device_type.cc:36]: one blank line

File: src/tir/transforms/bind_device_type.cc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      return GetRef<PrimExpr>(op);
+    }
+  }
+
+ public:
+  const VarNode* var_{nullptr};
+  int device_type_;
+};
+
+
```
Comment (zhiics) [inline src/tir/transforms/bind_device_type.cc:94]: one blank line

**Comment by tqchen** [comment]
cc @zhiics please take another look

**Pull request merged**